website/demos/ColumnsReordering.tsx

```diff
@@ -29,72 +28,81 @@ function createRows(): Row[] {
   return rows;
 }
 
-function createColumns(): Column<Row>[] {
-  return [
-    {
-      key: 'id',
-      name: 'ID',
-      width: 80
-    },
-    {
-      key: 'task',
-      name: 'Title',
-      resizable: true,
-      sortable: true
-    },
-    {
-      key: 'priority',
-      name: 'Priority',
-      resizable: true,
-      sortable: true
-    },
-    {
-      key: 'issueType',
-      name: 'Issue Type',
-      resizable: true,
-      sortable: true
-    },
-    {
-      key: 'complete',
-      name: '% Complete',
-      resizable: true,
-      sortable: true
-    }
-  ];
-}
+const columns: Column<Row>[] = [
+  {
+    key: 'id',
+    name: 'ID',
+    width: 80
+  },
+  {
+    key: 'task',
+    name: 'Title',
+    resizable: true,
+    sortable: true,
+    draggable: true
+  },
+  {
+    key: 'priority',
+    name: 'Priority',
+    resizable: true,
+    sortable: true,
+    draggable: true
+  },
+  {
+    key: 'issueType',
+    name: 'Issue Type',
+    resizable: true,
+    sortable: true,
+    draggable: true
+  },
+  {
+    key: 'complete',
+    name: '% Complete',
+    resizable: true,
+    sortable: true,
+    draggable: true
+  }
+];
 
 export default function ColumnsReordering({ direction }: Props) {
   const [rows] = useState(createRows);
-  const [columns, setColumns] = useState(createColumns);
+  const [columnsOrder, setColumnsOrder] = useState(() => {
+    const map = new Map<string, number>();
+    for (const [index, column] of columns.entries()) {
+      map.set(column.key, index);
+    }
+    return map;
+  });
   const [sortColumns, setSortColumns] = useState<readonly SortColumn[]>([]);
   const onSortColumnsChange = useCallback((sortColumns: SortColumn[]) => {
     setSortColumns(sortColumns.slice(-1));
   }, []);
 
-  const draggableColumns = useMemo(() => {
-    function renderHeaderCell(props: RenderHeaderCellProps<Row>) {
-      return <DraggableHeaderRenderer {...props} onColumnsReorder={handleColumnsReorder} />;
-    }
-
-    function handleColumnsReorder(sourceKey: string, targetKey: string) {
-      const sourceColumnIndex = columns.findIndex((c) => c.key === sourceKey);
-      const targetColumnIndex = columns.findIndex((c) => c.key === targetKey);
-      const reorderedColumns = [...columns];
-
-      reorderedColumns.splice(
-        targetColumnIndex,
-        0,
-        reorderedColumns.splice(sourceColumnIndex, 1)[0]
-      );
-
-      setColumns(reorderedColumns);
-    }
-
-    return columns.map((c) => {
-      if (c.key === 'id') return c;
-      return { ...c, renderHeaderCell };
+  const onColumnsReorder = useCallback((sourceKey: string, targetKey: string) => {
+    setColumnsOrder((columnsOrder) => {
+      const sourceColumnIndex = columnsOrder.get(sourceKey)!;
+      const targetColumnIndex = columnsOrder.get(targetKey)!;
```

Comment:
If we'd send the CalculatedColumn we could access `column.idx`

Reply:
https://github.com/adazzle/react-data-grid/blob/ba9f62117e74c61c4a8f0db825c80a167ba79603/src/HeaderCell.tsx#L222

We can add a function in `DataGrid.tsx` that can find the column based on the key. I think it is okay to have keys here but I can change it if you prefer

Reply:
https://github.com/adazzle/react-data-grid/blob/ba9f62117e74c61c4a8f0db825c80a167ba79603/src/HeaderCell.tsx#L222

We can add a function in `DataGrid.tsx` that can find the column based on the key. I think it is okay to have keys here but I can change it if you prefer